crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1902 @@
+                LeafNode {
+                    parent: Some(TreeIndex(0)),
+                    key: old_leaf.key,
+                    value: old_leaf.value,
+                    hash: old_leaf.hash,
+                },
+            ),
+            (
+                match side {
+                    Side::Left => left_index,
+                    Side::Right => right_index,
+                },
+                node,
+            ),
+        ];
+
+        for (index, node) in nodes {
+            let block = Block {
+                metadata: NodeMetadata {
+                    node_type: NodeType::Leaf,
+                    dirty: false,
+                },
+                node: Node::Leaf(node),
+            };
+
+            self.insert_entry_to_blob(index, &block)?;
+        }
+
+        Ok(())
+    }
+
+    fn insert_third_or_later(
+        &mut self,
+        mut node: LeafNode,
+        old_leaf: &LeafNode,
+        old_leaf_index: TreeIndex,
+        internal_node_hash: &Hash,
+        side: Side,
+    ) -> Result<(), Error> {
+        let new_leaf_index = self.get_new_index();
+        let new_internal_node_index = self.get_new_index();
+
+        node.parent = Some(new_internal_node_index);
+
+        let new_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node::Leaf(node),
+        };
+        self.insert_entry_to_blob(new_leaf_index, &new_leaf_block)?;
+
+        let (left_index, right_index) = match side {
+            Side::Left => (new_leaf_index, old_leaf_index),
+            Side::Right => (old_leaf_index, new_leaf_index),
+        };
+        let new_internal_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Internal,
+                dirty: false,
+            },
+            node: Node::Internal(InternalNode {
+                parent: old_leaf.parent,
+                left: left_index,
+                right: right_index,
+                hash: *internal_node_hash,
+            }),
+        };
+        self.insert_entry_to_blob(new_internal_node_index, &new_internal_block)?;
+
+        let Some(old_parent_index) = old_leaf.parent else {
+            panic!("root found when not expected")
+        };
+
+        self.update_parent(old_leaf_index, Some(new_internal_node_index))?;
+
+        let mut old_parent_block = self.get_block(old_parent_index)?;
+        if let Node::Internal(ref mut internal_node, ..) = old_parent_block.node {
+            if old_leaf_index == internal_node.left {
+                internal_node.left = new_internal_node_index;
+            } else if old_leaf_index == internal_node.right {
+                internal_node.right = new_internal_node_index;
+            } else {
+                panic!("child not a child of its parent");
+            }
+        } else {
+            panic!("expected internal node but found leaf");
```

Comment:
Are all these panics unreachable? You should use `unreachable!()` instead to indicate this. Otherwise, use `Error`.

Reply:
I think I do have some unreachable code somewhere that I wasn't able to get rid of via typing.  I don't think I would categorize either of these specifically as unreachable though.  Both of these asserts could come from data corruption elsewhere in the code I think.